File: search/array-based-search/fibonacci-search/cpp/fibonacci_search_test.cpp
#include <gtest/gtest.h>
#include <vector>

// Forward declaration
int fibonacciSearch(const std::vector<int>& arr, int target);

// Test: Target in middle
TEST(FibonacciSearchTest, TargetInMiddle) {
    std::vector<int> arr = {10, 20, 30, 40, 50, 60, 70, 80};
    EXPECT_EQ(fibonacciSearch(arr, 60), 5) << "Should find 60 at index 5";
}

// Test: Target at beginning
TEST(FibonacciSearchTest, TargetAtBeginning) {
    std::vector<int> arr = {1, 2, 3, 4, 5};
    EXPECT_EQ(fibonacciSearch(arr, 1), 0) << "Should find first element";
}

// Test: Target at end
TEST(FibonacciSearchTest, TargetAtEnd) {
    std::vector<int> arr = {1, 2, 3, 4, 5};
    EXPECT_EQ(fibonacciSearch(arr, 5), 4) << "Should find last element";
}

// Test: Not present
TEST(FibonacciSearchTest, TargetNotFound) {
    std::vector<int> arr = {1, 2, 3, 4, 5};
    EXPECT_EQ(fibonacciSearch(arr, 6), -1) << "Should return -1 for missing element";
}

// Test: Empty array
TEST(FibonacciSearchTest, EmptyArray) {
    std::vector<int> arr = {};
    EXPECT_EQ(fibonacciSearch(arr, 10), -1) << "Empty array should return -1";
}

// Test: Single element found
TEST(FibonacciSearchTest, SingleElementFound) {
    std::vector<int> arr = {5};
    EXPECT_EQ(fibonacciSearch(arr, 5), 0) << "Single element found → index 0";
}

// Test: Single element not found
TEST(FibonacciSearchTest, SingleElementNotFound) {
    std::vector<int> arr = {5};
    EXPECT_EQ(fibonacciSearch(arr, 10), -1) << "Single element not found → -1";
}

// Test: Small array with Fibonacci size (5)
TEST(FibonacciSearchTest, SmallArrayExactFibonacciSize) {
    std::vector<int> arr = {10, 20, 30, 40, 50}; // Size = 5 (Fibonacci number)
    EXPECT_EQ(fibonacciSearch(arr, 40), 3) << "Should find 40 at index 3";
}

// Test: Duplicate elements — should return valid index
TEST(FibonacciSearchTest, DuplicateElementsValidIndex) {
    std::vector<int> arr = {5, 5, 5, 5, 5};
    int result = fibonacciSearch(arr, 5);
    EXPECT_GE(result, 0) << "Result must be valid index";
    EXPECT_LT(result, 5) << "Result must be within bounds";
    EXPECT_EQ(arr[result], 5) << "Returned value must be 5";
}

// Test: Target smaller than first
TEST(FibonacciSearchTest, TargetSmallerThanFirst) {
    std::vector<int> arr = {10, 20, 30};
    EXPECT_EQ(fibonacciSearch(arr, 5), -1) << "Should return -1 if target < min";
}

// Test: Target larger than last
TEST(FibonacciSearchTest, TargetLargerThanLast) {
    std::vector<int> arr = {10, 20, 30};
    EXPECT_EQ(fibonacciSearch(arr, 40), -1) << "Should return -1 if target > max";
}

// Entry point for tests
int main(int argc, char** argv) {
    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}

File: search/array-based-search/fibonacci-search/java/FibonacciSearchTest.java
import org.junit.Test;
import static org.junit.Assert.*;

/**
 * Unit tests for the FibonacciSearch class.
 */
public class FibonacciSearchTest {

    @Test
    public void testTargetInMiddle() {
        int[] arr = {10, 20, 30, 40, 50, 60, 70, 80};
        assertEquals("Should find 60 at index 5", 5, FibonacciSearch.fibonacciSearch(arr, 60));
    }

    @Test
    public void testTargetAtBeginning() {
        int[] arr = {1, 2, 3, 4, 5};
        assertEquals("Should find first element", 0, FibonacciSearch.fibonacciSearch(arr, 1));
    }

    @Test
    public void testTargetAtEnd() {
        int[] arr = {1, 2, 3, 4, 5};
        assertEquals("Should find last element", 4, FibonacciSearch.fibonacciSearch(arr, 5));
    }

    @Test
    public void testTargetNotPresent() {
        int[] arr = {1, 2, 3, 4, 5};
        assertEquals("Should return -1 for missing element", -1, FibonacciSearch.fibonacciSearch(arr, 6));
    }

    @Test
    public void testEmptyArray() {
        int[] arr = {};
        assertEquals("Empty array should return -1", -1, FibonacciSearch.fibonacciSearch(arr, 10));
    }

    @Test
    public void testSingleElementFound() {
        int[] arr = {5};
        assertEquals("Single element found → index 0", 0, FibonacciSearch.fibonacciSearch(arr, 5));
    }

    @Test
    public void testSingleElementNotFound() {
        int[] arr = {5};
        assertEquals("Single element not found → -1", -1, FibonacciSearch.fibonacciSearch(arr, 10));
    }

    @Test
    public void testSmallArrayExactFibonacciSize() {
        int[] arr = {10, 20, 30, 40, 50}; // Size = 5 (Fibonacci number)
        assertEquals("Should find 40 at index 3", 3, FibonacciSearch.fibonacciSearch(arr, 40));
    }

    @Test
    public void testDuplicateElements_ReturnsValidIndex() {
        int[] arr = {5, 5, 5, 5, 5};
        int result = FibonacciSearch.fibonacciSearch(arr, 5);
        assertTrue("Result must be valid index", result >= 0 && result < 5);
        assertEquals("Returned value must be 5", 5, arr[result]);
    }

    @Test
    public void testTargetSmallerThanFirst() {
        int[] arr = {10, 20, 30};
        assertEquals("Should return -1 if target < min", -1, FibonacciSearch.fibonacciSearch(arr, 5));
    }

    @Test
    public void testTargetLargerThanLast() {
        int[] arr = {10, 20, 30};
        assertEquals("Should return -1 if target > max", -1, FibonacciSearch.fibonacciSearch(arr, 40));
    }
}